metadata-ingestion/src/datahub/ingestion/source/redshift/lineage.py

```diff
@@ -613,3 +648,143 @@ def get_lineage(
+            # Each part of sql has same transaction_id and start_time and unique sequence number
+            sql_parts[(row.transaction_id, row.start_time)][
+                row.sequence
+            ] = row.query_text
+
+        for row in temp_table_rows:
+            # filter out "create temp|temporary|#<table-name>" statements rows for the transaction
+            if not any(
+                row.query_text.lower().startswith(prefix)
+                for prefix in [
+                    RedshiftQuery.CREATE_TEMPORARY_TABLE_CLAUSE,
+                    RedshiftQuery.CREATE_TEMP_TABLE_CLAUSE,
+                    RedshiftQuery.CREATE_TEMPORARY_TABLE_HASH_CLAUSE,
+                ]
+            ):
+                continue
+
+            transaction_entry: Dict[int, str] = sql_parts[
+                (row.transaction_id, row.start_time)
+            ]
+
+            full_sql: str = RedshiftLineageExtractor._concat_sql_part(transaction_entry)
+
+            temp_tables.append(
+                TempTableRow(
+                    start_time=row.start_time,
+                    sequence=row.sequence,
+                    query_text=full_sql,
+                    transaction_id=row.transaction_id,
+                    process_identifier=row.process_identifier,
+                )
+            )
+
+        return temp_tables
+
+    def _add_permanent_datasets_recursively(
+        self,
+        db_name: str,
+        temp_table_names: List[str],
+        connection: redshift_connector.Connection,
+        permanent_lineage_datasets: List[LineageDataset],
+    ) -> None:
+
+        ddl_query: str = RedshiftQuery.temp_table_ddl_query(
+            tables=temp_table_names,
+            db_name=db_name,
+            start_time=self.config.start_time,
+            end_time=self.config.end_time,
+        )
+
+        logger.debug(f"Temporary table ddl query = {ddl_query}")
+
+        temp_table_rows: List[TempTableRow] = [
+            row
+            for row in RedshiftDataDictionary.get_temporary_rows(
+                conn=connection,
+                query=ddl_query,
+            )
+        ]
+
+        full_sql_temp_tables: List[
+            TempTableRow
+        ] = RedshiftLineageExtractor.form_full_sql(temp_table_rows)
+
+        transitive_temp_tables: List[str] = []
+
+        for temp_table in full_sql_temp_tables:
+            intermediate_l_datasets, _ = self._get_sources_from_query(
+                db_name=db_name,
+                query=temp_table.query_text,
+            )
+
+            # make sure lineage dataset should not contain a temp table
+            # if such dataset is present then add it to transitive_temp_tables to resolve it to original permanent table
+            for lineage_dataset in intermediate_l_datasets:
+
+                db, schema, table = split_qualified_table_name(lineage_dataset.urn)
+
+                # Check if table found is again a temp table
+                if not RedshiftDataDictionary.is_empty(
```

Comment:
This is quite expensive way to check if a table is a temp table. Ideally it should be done on the client side.

Reply:
removed